1) authorization

wmid1 opened the pull request:
Added authorization and routing betweet pages

wmid1 committed:
Added authorization and routing betweet pages

wmid1 committed:
small changes

shatalov-boris commented on .eslintrc.js:
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+    extends: ['eslint-config-purrweb-react'],
+  };
```
@wmid1 
Нужно всегда добавлять пустую стрку последней в файл - https://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-file
В VS code есть опция для автоматичесчкого добавления пустой строки при сохранение.
Нужно включить `insertFinalNewline` в настройках

shatalov-boris commented on src/containers/SignInPopUp.js:
```diff
@@ -0,0 +1,165 @@
+  authChange: auth => dispatch(authChange(auth)),
+});
+
+SignInAuth.propTypes = {
```
@wmid1 
Объявлять пропсы лучше через статичную функцию сверху компонента - https://stackoverflow.com/a/39683633/8193955

shatalov-boris commented on src/containers/SignInPopUp.js:
```diff
@@ -0,0 +1,165 @@
+
+  axiosPost = () => {
+    if (this.state.email.length > 3 && this.state.password.length > 7) {
+      axios
```
@wmid1 
Несколько моментов:
1) Запросы к апи нужно вынести в отдельный файл `api.js` и объявить в виде функций:
```
export function fetchSignIn(email, password) {
  axios.post('https://floating-atoll-63112.herokuapp.com/api/auth/sign_in', {
    email: email,
    password: password,
  });
}
````
Затем ты можешь просто вызывать 
```
fetchSignIn(email, password).then(() => {}).catch(() => {});
```
2)
Адрес апи надо хранить в отдельной переменной. Это упрощает работу, если он вдруг изменится -н ужно будет поменять только одно место вместо поиска всех упоминаний в коде.
```
const AUTH_API_URL = 'https://floating-atoll-63112.herokuapp.com/api/auth';

export function fetchSignIn(email, password) {
  axios.post(`${AUTH_API_URL}/sign_in`, {
    email: email,
    password: password,
  });
}
```

3) Называть функции нужно, чтобы было понятно, что они делают `axiosPost` говорит о том, что просто происходит post запрос всё. Если это запрос для логина, нужно назвать функцию signIn

shatalov-boris commented on src/containers/SignInPopUp.js:
```diff
@@ -0,0 +1,165 @@
+          this.props.authChange(true);
+        })
+        .catch(error => {
+          console.log(error);
```
@wmid1 
 Здесь нужно использовать `toastre.error`, чтобы отобразить ошибку пользователю

shatalov-boris commented on src/containers/SignInPopUp.js:
```diff
@@ -0,0 +1,165 @@
+                    <span className="error-message">Too short. Use at least 8 characters</span>
+                  </div>
+                  <button className="btn btn-blue btn-with-icon">
+                    <span className="button-content" onClick={this.axiosPost}>
```
@wmid1 
Отсюда нужно убрать `onClick={this.axiosPost}`. 
Вместо этого нужно обернуть все инпуты в тег `<form>` и реагировать на сабмит формы.
`form onSubmit={this.axiosPost}`
Таким образом форма обработаетсядаже если юзер нажмёт enter.

shatalov-boris commented on src/containers/SignInPopUp.js:
```diff
@@ -0,0 +1,165 @@
+    }
+  };
+
+  fName = event => {
```
@wmid1 
Имя функция должно содержать в имени какое-то действие. В этом случае подходящим будет `handleFName`.
Но вообще, эту и другие фунции обработки инпутов нужно заменить одной функцией `handleInputChange`
В ней ты будешь смотреть имя инпута и по нему записывать в стейт.

shatalov-boris commented on src/containers/logInPopUp.js:
```diff
@@ -0,0 +1,138 @@
+import { authentication, authChange } from '../actions/actions';
+import '../styles/home-styles.css';
+
+class LogInAuth extends Component {
```
@wmid1 Имя файла должно совпадать с именем компонента

shatalov-boris reviewed (changes requested):
@wmid1 
Общие замечания:

1) Форма логина и регистрации не должна показывать ошибки валидации сразу при открытии.
2) После сабмита формы нужно отключать возможность нажать на кнопку пока зарос не завершится. Сейчас можно послать кучу запросов просто кликая по кнопке.
3) Посмотреть все теги `<button>`, по умолчанию они имеют `type=submit`, то есть пытаются обработать форму. И это может привести к нелжиданным сабмитам. Нужно в те кнопки, которые не связаны с формой прописывать `type=button`.

wmid1 committed:
corrected errors

shatalov-boris commented on src/api.js:
```diff
@@ -0,0 +1,29 @@
+import axios from 'axios';
+
+const AUTH_API_URL = 'https://floating-atoll-63112.herokuapp.com';
```
@wmid1 
Тут надо использовать именно 
```
const AUTH_API_URL = 'https://floating-atoll-63112.herokuapp.com/api/auth';
```
И потом ты можешь не прописывать `api/auth` в урлах.

Даже можно вот так:
```
const BASE_API_URL = 'https://floating-atoll-63112.herokuapp.com';
const AUTH_API_URL = `${BASE_API_URL}/api/auth`;
```

shatalov-boris commented on src/api.js:
```diff
@@ -0,0 +1,29 @@
+  });
+}
+
+export function validation(userDate) {
```
@wmid1 
Тут опечатка, не `userDate`, а `userData`

shatalov-boris commented on src/containers/RegisterPopUp.js:
```diff
@@ -1,64 +1,96 @@
+import toastr from 'toastr';
 import { authentication, authChange } from '../actions/actions';
+import { fetchRegister } from '../api';
+import '../../node_modules/toastr/build/toastr.css';
```
@wmid1 
Файлы из нод модулей нудно подключать так:
`import '~toastr/toastr.scss';`
То есть использовать `~` вместо полного пути к модулю

shatalov-boris commented on src/containers/RegisterPopUp.js:
```diff
@@ -83,55 +115,61 @@ class SignInAuth extends Component {
+                    <input
+                      type="submit"
+                      value="➡ START NOW"
+                      ng-click="$ctrl.login(sessionsForm)"
```
@wmid1 
Не нужно использовать `ng-` подобные аттрибуты, они были в вёрстке, но сейчас они бессмысленны. Они нужны для другого фрейморка Angular, которые использовался в проекте.

shatalov-boris commented on src/containers/logInPopUp.js:
```diff
@@ -1,52 +1,78 @@
-  emailSend = event => {
-    this.setState({ email: event.target.value });
+  handleInputChange = event => {
+    const nameInput = event.target.name;
```
@wmid1 
Здесь нужно использовать деструктуризацию
```
const { target: { value, name } } = event;
```
Вместо того, чтобы прописывать каждый раз полный путь к аттрибуту типо `event.target.value`

shatalov-boris commented on src/pages/App.js:
```diff
@@ -62,22 +56,21 @@ class App extends Component {
               Search
             </a>
-            <div className="item-text" onClick={() => this.logOut()}>
+            <div className="item-text" onClick={() => this.logExit()}>
```
@wmid1 
Ну нужно определать функцию вот так:
```onClick={() => this.logExit()}```
В этом случае функция будет создаваться каждый раз на ререндер. Это не влияет на работу программы, но засоряет память. Не серьёзно, но лучше так не делать. На больших проектах с кучей таких моментов может проседать производительность.

Нужно так:
```
onClick={this.logExit}
```
Он создаст один обработчик и всё.

Это встречается ещё в нескольких местах.

shatalov-boris commented on src/containers/RegisterPopUp.js:
```diff
@@ -1,64 +1,96 @@
         .catch(error => {
+          toastr.error('Email already in use');
+          this.setState({ onSubmit: true });
           console.log(error);
```
@wmid1 
Не нужно коммитить `console.log`

shatalov-boris commented on src/containers/RegisterPopUp.js:
```diff
@@ -1,64 +1,96 @@
+          });
         })
         .catch(error => {
+          toastr.error('Email already in use');
```
@wmid1 
Здесь нужно брать текст ошибки из объекта `error`. Потому что ошибка может быть и не связана с емаилом.

shatalov-boris commented on src/containers/RegisterPopUp.js:
```diff
@@ -1,64 +1,96 @@
-
-  passSend = event => {
-    this.setState({ password: event.target.value });
+  waitSubmitForm = event => {
```
@wmid1 
Эта функция вообще не нужна.
Вместо этого нужно добавить пропс `disabled` к кнопке сабмита и менять true/false. При true она перестанет реагировать на клики и изменится курсор.

shatalov-boris commented on src/containers/RegisterPopUp.js:
```diff
@@ -83,55 +115,61 @@ class SignInAuth extends Component {
-                  </button>
-                </div>
+                  <div className="form-group">
+                    <input
```
@wmid1 
Для сабмита надо использовать `<button type="submit" />`, а не `<input type="submit" />`
Разницы функционала нет, но кнопка имеет больше возможностей для взаимодествия.

shatalov-boris commented on src/containers/RegisterPopUp.js:
```diff
@@ -1,64 +1,96 @@
   close = () => {
     this.props.modalClose(false);
-    this.setState({ value: '', email: '', first_name: '', last_name: '', password: '' });
+    this.setState({
```
@wmid1 
Здесь не нужно обнулять state при закрытии, потому что компонент в этом случае делает unmount и при новом открытии автоматически возмётся начальное состояние.

shatalov-boris replied:
@wmid1 Здесь не исправлено
>Здесь не нужно обнулять state при закрытии, потому что компонент в этом случае делает unmount и при новом открытии автоматически возмётся начальное состояние.

shatalov-boris commented:
@wmid1 
Момент по коммитам - последний коммит `corrected errors`, а до этого был `small changes`.
1) Нужно писать с большой буквы и начинать коммит с глагола. Также использовать настоящее время. То есть первый коммит бы был `Correct errors`.
2) Они не несут никакой полезной информации о том, что же конкретно было сделано. Просто абстрактные исправления. Если потом понадобится посмотреть историю коммитов, чтобы найти ошибку, то по таким сообщениям будет не понятно, что включено в коммит.  
Они были бы приемлемыми, если бы у них были комментарии. Комментарий к коммиту пишутся в самом коммите, но через пустую строку, вот так:
```
Correct errors

1) Fix ...
2) Update ...
3) Remove ...
```
Заголовок коммита должен быть коротким, а вот комментария может быть большим, где всё подробно расписано.

Но лучше избегать такого большого расписывания. Лучше коммитить небольшими кусками.

Проблема в том, что ты написал `corrected errors`, потому что у тебя было слишком много изменения и ты не мог описать их одним коротким предложением. Если такое случается, то это говорит о том, что нужно разбивать коммит по маленьким связанным кускам.

 Можешь ещё вот [это](https://habr.com/ru/company/Voximplant/blog/276695/) почитать

shatalov-boris commented:
@wmid1 
В данном проекте тебе не нужно будет использовать `redux-persist` пакет. Нужно его удалить, потому что здесь нет данных, которые нужно хранить локально, чтобы потом их подгружать. Все данные приходят через апи. Ты только хранишь токен в localStorage.

shatalov-boris commented on src/containers/RegisterPopUp.js:
```diff
@@ -1,64 +1,96 @@
+    const nameInput = event.target.name;
+    this.setState({ [nameInput]: event.target.value });
+    if (event.target.name === 'password') {
+      if (event.target.value.length < 8) {
```
@wmid1 
Такое
```
     if (event.target.value.length < 8) {
        this.setState({ passwordValid: false });
      } else {
        this.setState({ passwordValid: true });
      }
```
Можно заменить вот этим
```
this.setState({ passwordValid: event.target.value.length < 8 });
```
То есть здесь не нужный if else

shatalov-boris commented on src/reducers/configureStore.js:
```diff
@@ -0,0 +1,24 @@
+
+export const history = createBrowserHistory();
+
+const persistConfig = {
```
@wmid1 
В данном проекте тебе не нужно будет использовать `redux-persist` пакет. Нужно его удалить, потому что здесь нет данных, которые нужно хранить локально, чтобы потом их подгружать. Все данные приходят через апи. Ты только хранишь токен в localStorage.

wmid1 committed:
Fix link

wmid1 committed:
Transfer import toaster.css to styles

wmid1 committed:
Delete ng- tags

wmid1 committed:
Use destructuring in handleInputChange

wmid1 committed:
Show error in toastr

wmid1 committed:
Delete func waitSubmitForm

1) Instead input to button
2) Add prop disabled to button

wmid1 committed:
Remove unnecessary if-else

wmid1 committed:
Delete redux-persistor

1) Use localStorage to keep access-token and other user info
2) Delete all console.log()

shatalov-boris commented on src/index.js:
```diff
@@ -1,12 +1,28 @@
+/* eslint-disable no-undef */
 import React from 'react';
 import ReactDOM from 'react-dom';
 import './index.css';
```
@wmid1 
Относительный import идёт после абсолютного

shatalov-boris commented on src/index.js:
```diff
@@ -1,12 +1,28 @@
+  <Provider store={store}>
+    <ConnectedRouter history={history}>
+      <Switch>
+        <Route exact path="/" component={App} />
```
@wmid1 
Так как это страница, то лучше назвать компонент Home, а не App.
Страница App - это что-то странное)

shatalov-boris commented on src/pages/App.js:
```diff
@@ -0,0 +1,123 @@
+  };
+
+  state = {
+    modalLogIn: false,
```
@wmid1 
Небольшой момент:
log in и sign in - это синонимы, так же register и sign up - это тоже синонимы. 
В нашем случае вместо sign in нужно sign up.
И как правило используют связку sign in - sign up или log in - register. Но не смешивают.

Теперь что сделать:
1) Нужно назвать `isSigninModalOpen ` и `isSignupModalOpen ` вместо `modalLogIn` и `modalSignIn` 
2) Функции `modalLogIn` и `modalSignIn` переименовать в `openSigninModal` и `openSignupModal`. И теперь им не надо передавать аргумент, они всегда будет ставить true в стейт.
3) `modalClose` тоже разбить на 2 функции для каждой модалки как в пункте 2.

shatalov-boris commented on src/pages/App.js:
```diff
@@ -0,0 +1,123 @@
+          <h2>
+            <b>Welcome!</b>
+          </h2>
+          <LogInPopUp modalLogIn={this.state.modalLogIn} modalClose={this.modalClose} />
```
@wmid1 
Вот так не надо делать `modalClose={this.modalClose}` и потом внутри компонента проверять и возвращать null, если это false.
Рендер компонента надо по возможности предотвращать в родительском компоненте:
```
{this.modalClose && <LogInPopUp modalLogIn={this.state.modalLogIn} />}
```
Он отрендерить модалку только если `modalClose` - true

shatalov-boris commented on src/pages/App.js:
```diff
@@ -0,0 +1,123 @@
+    logOut().then(() => {
+      this.props.authentication({});
+    });
+    // eslint-disable-next-line no-undef
```
@wmid1 
Вместо того, чтобы постоянно прописывать игнорирование для переменной `locaStorage` - можно просто прописать игнорирование в .eslintrc.js
```
module.exports = {
  extends: ['eslint-config-purrweb-react'],
  globals: {
    'localStorage': false,
  },
};
```
И больше не нужно отключать проверку комментариями

shatalov-boris commented on src/pages/App.js:
```diff
@@ -0,0 +1,123 @@
+
+  validateToken = () => {
+    validation().catch(() => {
+      toastr.error('Authorized users only.');
```
@wmid1 
Здесь, помимо вывода ошибки, ещё нужно очишать данные из localStorage, чтобы сделать юзера гостевым. Так как у него не валидный токен, то и данные, которые у нас хранятся уже не валидны.

shatalov-boris commented on src/pages/App.js:
```diff
@@ -0,0 +1,123 @@
+    userInfo: store.userInfo,
+  };
+}
+const mapDispatchToProps = dispatch => ({
```
@wmid1 
Нужно не забыть оставлять пустую строку

shatalov-boris commented on src/pages/App.js:
```diff
@@ -0,0 +1,123 @@
+  render() {
+    // eslint-disable-next-line no-undef
+    if (localStorage.getItem('auth')) {
+      this.validateToken();
```
@wmid1 
Эту проверку надо вынести в componentDidMount

shatalov-boris commented on src/pages/App.js:
```diff
@@ -0,0 +1,123 @@
+          <LogInPopUp modalLogIn={this.state.modalLogIn} modalClose={this.modalClose} />
+          <RegisterPopUp modalSignIn={this.state.modalSignIn} modalClose={this.modalClose} />
+          <div className="nav-btns flexbox justify-content-center">
+            <a href="Skills" className="item-text">
```
@wmid1 
Для перехода по роутам надо вместо тега `<a>` использовать компонент `<Link>` из пакета роутера - https://reacttraining.com/react-router/web/api/Link

shatalov-boris commented on src/containers/RegisterPopUp.js:
```diff
@@ -0,0 +1,178 @@
+          this.props.modalClose(false);
+        })
+        .catch(error => {
+          toastr.error(error);
```
@wmid1 
`error`, который здесь приходит - это объект, его не надо сразу передавать в `toastr.error`.
`toastr.error` ожидает строку и сейчас он выводит ошибку не понятную простому пользователю.
`error` выглядит вот так:
```
{"status":"error","data":{"id":null,"provider":"email","uid":"","first_name":"чмсчячс","last_name":"ячсячс","image":{"url":"https://s3.amazonaws.com/wll-community-production/images/no-avatar.png"},"email":"shatalov.bs@purrweb.com","created_at":null,"updated_at":null,"city":null,"country":null,"address":null,"zipcode":null,"mobile":null,"name":null,"nickname":null,"image_original":{"url":"https://s3.amazonaws.com/wll-community-production/images/no-avatar.png"},"experience":null,"availability":null,"place_to_work":null,"price":null,"cover_letter":null,"last_seen_at":null,"user_skill_categories_count":0,"saved_count":0,"profile_viewed":0,"puid":null,"puid_created":false,"rate":null,"skill_test_rate":null,"date_of_birthday":null,"wallet_id":null,"skill_test_send":0},"errors":{"email":["already in use"],"full_messages":["Email already in use"]}}
```
<img width="1153" alt="screenshot 2019-01-21 at 18 19 00" src="https://user-images.githubusercontent.com/4655654/51474314-145f9e00-1da9-11e9-99e7-c7d14f0c87f0.png">

Тебе надо показывать пользователю `full_messages`, то есть будет `toastr.error(error.errors.full_messages[0])`

shatalov-boris commented on src/containers/logInPopUp.js:
```diff
@@ -0,0 +1,149 @@
+          this.props.modalClose(false);
+        })
+        .catch(() => {
+          toastr.error('Invalid login credentials. Please try again.');
```
@wmid1 
Здесь нужно выводить ошибку тоже через приходящий `error`

shatalov-boris reviewed (changes requested):
@wmid1 
Общее:
1) Нужно создать компонент `<Header>` и вынести в него ссылки, так как сейчас происходит их дублирование.
2) 
Использовать деструктуразацию внутри render (сейчас там много вызовов типо `this.state.blabla`)
Также в `postRegisterForm` и `postLogInForm` - сейчас стейт нормально, а пропсы прописываются полностью. Их тоже нужно через деструктуризацию:
```
    const {
      props: {},
      state: {}
    } = this;
```
3) Переименовать папку `containers` в`components`

wmid1 committed:
Rename shared folder to components

wmid1 committed:
Add to ignore localStorage and document

wmid1 committed:
Rename page App to Home

wmid1 committed:
Prevent of unnecessary render in the child component

1) Rename any modal functions and states
2) Delete unnecessary PropTypes and if-else constructios

wmid1 committed:
New component with navigation links

1) Use Link and NavLink to switch between pages
2) Update styles for links
3) Use destructuring in LogInPopUp and RegisterPopUp

wmid1 committed:
Show full error message in toastr

wmid1 committed:
Delete func logExit

wmid1 committed:
Fix toastr error

wmid1 committed:
Rename logInPopUp.js to LogInPopUp.js
